File: react_client/src/features/media/mediaSchema.ts
const MediaType = {
    IMAGE: "image",
    VIDEO: "video",
} as const

type MediaType = typeof MediaType[keyof typeof MediaType]

const EntityType = {
    PRODUCT: "product",
    SALE: "sale",
} as const

type EntityType = typeof EntityType[keyof typeof EntityType]

type MediaDetail = {
    id: number
    mediaType: MediaType
    mediaUrl: string
    entityId: number
    entityType: EntityType
    altText: string | null
    sortOrder: number | null
}

/*
ROUTES
GET /products/:id/media
POST /products/media
GET /sales/:id/media
POST /sales/media

GET /products/export
GET /sales/export
POST /products/import
POST /sales/import
*/

// Products
type GetProductMediaResponse = {
    media: MediaDetail[]
}

type RetrieveProductsMediaRequest = {
    productIds: number[]
}

type RetrieveProductsMediaResponse = {
    media: Record<number, MediaDetail>
}

// Sales
type GetSaleMediaResponse = {
    media: MediaDetail[]
}

type RetrieveSalesMediaRequest = {
    saleIds: number[]
}

type RetrieveSalesMediaResponse = {
    media: Record<number, MediaDetail>
}

// CSV Import Export
type CsvImportRequest = {
    file: File
}

type CsvImportResponse = {
    imported: number
    updated: number
    failed: number
    errors: {
        row: number
        field: string | null
        message: string
    }[]
}

export type {
    MediaDetail,
    GetProductMediaResponse,
    RetrieveProductsMediaRequest,
    RetrieveProductsMediaResponse,
    GetSaleMediaResponse,
    RetrieveSalesMediaRequest,
    RetrieveSalesMediaResponse,
    CsvImportRequest,
    CsvImportResponse,
}

export { 
    MediaType, 
    EntityType 
}

File: fastapi_server/app/features/media/media_schema.py
from typing import Literal

from app.api.api_model import APIModel
from app.db.models.media import Media


MediaType = Literal["image", "video"]
EntityType = Literal["product", "sale"]


class MediaDetail(APIModel):
    id: int
    media_type: MediaType
    media_url: str
    entity_id: int
    entity_type: EntityType
    alt_text: str | None
    sort_order: int | None

    @staticmethod
    def from_Media(media: Media, media_url: str):
        return MediaDetail(
            id=media.id,
            media_type=media.type,
            media_url=media_url,
            entity_id=media.entity_id,
            entity_type=media.entity_type,
            alt_text=media.alt_text,
            sort_order=media.sort_order,
        )


"""
ROUTES
GET /products/:id/media
POST /products/media
GET /sales/:id/media
POST /sales/media
GET /media/:entity_type/:file_name

GET /products/export
GET /sales/export
POST /products/import
POST /sales/import
"""


# Products
class GetProductMediaResponse(APIModel):
    media: list[MediaDetail]


class RetrieveProductsMediaRequest(APIModel):
    product_ids: list[int]


class RetrieveProductsMediaResponse(APIModel):
    media: dict[int, MediaDetail]


# Sales
class GetSaleMediaResponse(APIModel):
    media: list[MediaDetail]


class RetrieveSalesMediaRequest(APIModel):
    sale_ids: list[int]


class RetrieveSalesMediaResponse(APIModel):
    media: dict[int, MediaDetail]


class CsvImportError(APIModel):
    row: int
    field: str | None = None
    message: str


class CsvImportResponse(APIModel):
    imported: int
    updated: int
    failed: int
    errors: list[CsvImportError]
